### DISCUSSION TITLE
Python tasks

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#9834] [Show and tell]
A couple of simple tasks to enable faster python execution. Add these to your `tasks.json` file. Swap out `python3` for whatever interpreter/alias you are using. Use `tasks: spawn` in the command palette to pick a task and then relaunch the same task via `opt-t` (`task: rerun`).

Executing code

```json
{
  "label": "Run python selection",
  "command": "python3",
  "args": ["-c", "\"$ZED_SELECTED_TEXT\""],
  "use_new_terminal": false
},
```
```json
{
  "label": "Run python file",
  "command": "python3",
  "args": ["$ZED_FILE"],
  "use_new_terminal": false
}
```

Managing venvs (utilizing [`uv`](https://github.com/astral-sh/uv))

```json
{
  "label": "Create venv",
  "command": "uv",
  "args": ["venv", "$ZED_WORKTREE_ROOT/.venv"],
  "use_new_terminal": false
},
```

```json
{
  "label": "Delete venv",
  "command": "rm",
  "args": ["-rf", "$ZED_WORKTREE_ROOT/.venv"],
  "use_new_terminal": false
}
```

<img width="631" alt="SCR-20240326-pagw" src="https://github.com/zed-industries/zed/assets/19867440/8ba2eb61-8387-4a34-a1c6-3779631217da">

For more general info on tasks, see [the docs](https://zed.dev/docs/tasks).

**Comment by KiraMelnikov** [comment]
i dont understand how it works (
for me it doesnt works (
![2024-04-04 15 59 11](https://github.com/zed-industries/zed/assets/120144794/70b9c85d-9122-4fa0-9f1a-22a5d7a14b7f)

**Reply by bennetbo** [reply]
You can specify multiple tasks in the tasks.json file, therefore you need to wrap it in a JSON array.

Works for me if I do it like this:

```json
[
  {
    "label": "Run python file",
    "command": "python3",
    "args": ["$ZED_FILE"],
    "use_new_terminal": false
  }
]
```

**Reply by KiraMelnikov** [reply]
yes, i`ve tried to wrap it in array firstly 
but i don`t understand for next to step what i have to do
I`ve pressed opt-t or opt-shift-t into python file but nothing to happened ...

**Reply by bennetbo** [reply]
You need to open the command palette (Cmd+Shift+P if you are using the default keybindings) and select `tasks: spawn`, then select the task and press enter. See here:

https://github.com/zed-industries/zed/assets/53836821/257235a3-cc1e-4b87-89af-2ece4ce1f59c

**Reply by KiraMelnikov** [reply]
Thank you so much for a video!))

I assumed it, but i havent the same menu 0_0

<img width="534" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/4d67c18c-6278-4a27-8cd7-6b75ef837fd0">
<img width="487" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/7738e750-56b2-434c-b56e-a9b27c492746">

**Reply by bennetbo** [reply]
And what happens when you run `open tasks` in the command palette? If you see the python task in that file I am not sure what's going wrong. You could check the logs for any errors via the command palette (zed: open log)

**Reply by KiraMelnikov** [reply]
i understood whats going on )
im created a tasks.json into project dir 
if i press cmd+shift+p 
ive seen this "Zed: open tasks"
Now i understand what i need to do
to add new tasks into .config/zed dir )
Thx for help and supporting to me!)
<img width="613" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/48cb2167-432b-4def-ba0e-feba12afac24">

**Reply by JosephTLyons** [reply]
Sorry for not providing more details - we have some documentation here:

- https://zed.dev/docs/tasks

**Reply by KiraMelnikov** [reply]
Thank you Joseph)
I had to check this earlier

**Reply by distrair** [reply]
Hi, I pasted your code in my tasks.json and for some reason, it doesn't appear in any of the .py files, but it appears in any other file type. How can I fix this?

**Comment by Tom-Boscher** [comment]
I wrote a Nu task to initialize a Python project. It create a `pyrightconfig.json`, a `.venv` and install all Python package in `requirements.txt` or create the file if missing. You have to have Nushell as login shell.
Sorry for the long command.

```
  {
    "label": "Setup Python project",
    "command": "if ('pyrightconfig.json' | path exists) == false {{'venvPath': '.', 'venv': '.venv'} | to json o> pyrightconfig.json} ; if ('.venv' | path exists) == false {uv venv} ; if ('requirements.txt' | path exists) == false {touch requirements.txt} ; if (cat requirements.txt | is-not-empty) {uv pip install -r requirements.txt}",
    "reveal": "never",
    "use_new_terminal": false
  }
  ```

**Reply by KiraMelnikov** [reply]
zsh:1: parse error near `}'

**Reply by Tom-Boscher** [reply]
@KiraMelnikov You have to use Nushell instead of Zsh to run this task as the command is wrote in Nu language

**Reply by KiraMelnikov** [reply]
@Tom-Boscher Thx! maybe you have know how can i to run python code via opt-t (or opt-shift-t) together with some venv when i already created ?
i already have a "pyrightconfig.json"
i already run my code via opt-t but it runs into main os path without .some_env 

I can activate source venv, yes and then to use pip install or enter "python3 name.py" and so on. 
But it's not very convenient (

Thx in anvance!

**Reply by Tom-Boscher** [reply]
@KiraMelnikov If you use task to run a python file it doesn't use your virtual environment?

**Reply by KiraMelnikov** [reply]
@Tom-Boscher Exactly so 
But if i install pip into main os path then its works 
If i activate env using terminal, RUN FILE not see packages in my .API because look at main os packages...

<img width="485" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/264d8968-0647-4079-afcd-e4e23d6879c8">
<img width="346" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/3cfff67f-9fad-429c-9ca8-70450809a56e">
<img width="437" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/ec273912-cbb1-48dd-b2a4-e789ddfac868">
<img width="514" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/e01da97a-8196-498e-9532-01bad4bb9eec">
<img width="565" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/a7cb3ea8-b960-4a83-bd74-7d8fe4f65936">

**Reply by Tom-Boscher** [reply]
Zed will looks (by default) for `.env`, `.venv`, `env`, `venv`
You can add other folder names in your configuration file (I'm not on my computer now so I can't tell you the exact config name)

**Reply by KiraMelnikov** [reply]
@Tom-Boscher 
I have create a new .venv then installed test_lib and my code is running
But the packages are installed into main folder and the same time into .venv 
When i uninstall packages from main folder so .venv become is empty too ... ive no idea
<img width="347" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/b73c2587-ea1a-433b-9b8f-0f646ee39dc6">
<img width="297" alt="image" src="https://github.com/zed-industries/zed/assets/120144794/8261d5ab-7bf2-4fdb-9a18-ed7abef3a432">

**Reply by kacperkwapisz** [reply]
I made a Zsh version of this command. No need to install/setup anything new. Enjoy 👀

(utilizing [uv](https://github.com/astral-sh/uv)):
```json
{
    "label": "Setup Python project",
    "command": "[ ! -f \"pyrightconfig.json\" ] && echo '{\"venvPath\": \".\", \"venv\": \".venv\"}' > pyrightconfig.json; [ ! -d \".venv\" ] && uv venv; [ ! -f \"requirements.txt\" ] && touch requirements.txt; [ -s \"requirements.txt\" ] && uv pip install -r requirements.txt",
    "reveal": "never",
    "use_new_terminal": false
}
  ```
  
(utilizing [pip](https://pypi.org/project/pip/)):
```json
{
  "label": "Setup Python project",
  "command": "[ ! -f \"pyrightconfig.json\" ] && echo '{\"venvPath\": \".\", \"venv\": \".venv\"}' > pyrightconfig.json; [ ! -d \".venv\" ] && python3 -m venv .venv; [ ! -f \"requirements.txt\" ] && touch requirements.txt; [ -s \"requirements.txt\" ] && .venv/bin/pip install -r requirements.txt",
  "reveal": "never",
  "use_new_terminal": false
}
```

**Comment by wmstack** [comment]
I am trying this with a simple java file, swapping python3 with java. It works in the terminal but the same command doesn't work with the task runner. Even though there is a `"$ZED_FILE"` argument, it still outputs as if `java` is called standalone.

The problem seems to be that with `java` files, "$ZED_FILE" is not outputting the correct name. I made a dummy task that does `echo` to the argument "$ZED_FILE" and it works for some files I have tried but not others. It doesn't work with txt files or java files, but seems to work with markdown and python files.

**Comment by HamzaFarhan** [comment]
Some more useful tasks:

```json
[
  // run python file using uv
  {
    "label": "Run File: $ZED_FILENAME",
    "command": "uv run $ZED_FILENAME"
  },
  // run selected code using uv
  {
    "label": "Run Code",
    "command": "cat <<'EOF' > temp_run.py\n$ZED_SELECTED_TEXT\nEOF\nuv run temp_run.py && rm temp_run.py"
  },
  // push config to github
  {
    "label": "Push Config",
    "command": "cd ~/.config/zed && git add . && git commit -m 'Update Config' && git push"
  }
]
```
The keymap:
```json
[
  {
    "context": "Workspace",
    "bindings": {
      "f4": ["task::Spawn", { "task_name": "Run Code" }],
      "f5": ["task::Spawn", { "task_name": "Run File: $ZED_FILENAME" }],
      "f7": "task::Spawn",
      "f6": "task::Rerun"
    }
  }
]
```

**Comment by injust** [comment]
This gets you pretty close to VS Code's "Python: Run Python File" command.

Keybinding:
```json
{
  "context": "Editor && mode == full && !(ContextEditor > Editor) || Terminal",
  "use_key_equivalents": true,
  "bindings": {
    "cmd-r": ["task::Spawn", { "task_name": "Python: $ZED_STEM" }]
  }
}
```

Task:
```json
{
  "label": "Python: $ZED_STEM",
  "command": "python3",
  "args": ["$ZED_FILE"],
  "allow_concurrent_runs": true
}
```

**Reply by injust** [reply]
This is what I use in my keymap now:

```json
{
  "context": "Editor && mode == full && !(ContextEditor > Editor) && extension == py",
  "use_key_equivalents": true,
  "bindings": {
    "cmd-r": ["task::Spawn", { "task_name": "run '$ZED_FILE'" }]
  }
}
```

You don't even need to create a task, since Zed's Python support already creates the `run '$ZED_FILE'` task.

This has the added benefit of automatically using the configured venv.

**Comment by eyayaw** [comment]
If you want to run with a venv python, this may be useful.

```json
 {
    "label": "Python: Run File",
    "command": "source $ZED_WORKTREE_ROOT/.venv/bin/activate; python3 || python3",
    "args": ["$ZED_FILE"],
    "use_new_terminal": false
  },
  {
    "label": "Python: Run Selection",
    "command": "source $ZED_WORKTREE_ROOT/.venv/bin/activate; python3 || python3",
    "args": ["-c", "\"$ZED_SELECTED_TEXT\""],
    "use_new_terminal": false
  }

```
